src/features/ci-cd/hooks/service-installs.ts

```diff
@@ -155,27 +197,87 @@ hooks.addPureHook('POST', 'resin', 'device', {
 });
 
 hooks.addPureHook('PATCH', 'resin', 'device', {
-	POSTRUN: async (args) => {
+	PRERUN: async (args) => {
 		const affectedIds = args.request.affectedIds!;
 
-		// We need to delete all service_install resources for the current device and
+		// We need to delete all service_install resources for the current device / app and
 		// create new ones for the new application (if the device is moving application)
 		if (
 			args.request.values.belongs_to__application != null &&
 			affectedIds.length !== 0
 		) {
-			await args.api.delete({
-				resource: 'service_install',
+			const deleteHandler = (
+				api: sbvrUtils.PinejsClient,
+				deviceIds: number[],
+				serviceIds: number[],
+			) =>
+				api.delete({
+					resource: 'service_install',
+					options: {
+						$filter: {
+							device: { $in: deviceIds },
+							installs__service: { $in: serviceIds },
+						},
+					},
+				});
+
+			const createHandler = async (
+				api: sbvrUtils.PinejsClient,
+				deviceIds: number[],
+				serviceIds: number[],
+			) => {
+				await Promise.all(
+					deviceIds.map(async (deviceId) => {
+						await Promise.all(
+							serviceIds.map(async (serviceId) => {
+								// Create a service_install for this pair of service and device
+								await api.post({
+									resource: 'service_install',
+									body: {
+										device: deviceId,
+										installs__service: serviceId,
+									},
+									options: { returnResource: false },
+								});
+							}),
+						);
+					}),
+				);
+			};
+
+			// get devices that are currently on a different app and delete those service installs
+			const devices = (await args.api.get({
+				resource: 'device',
 				options: {
+					$select: ['id', 'belongs_to__application'],
 					$filter: {
-						device: { $in: affectedIds },
+						id: { $in: affectedIds },
 					},
 				},
-			});
-			await createAppServiceInstalls(
-				args.api,
-				args.request.values.belongs_to__application,
-				affectedIds,
+			})) as Array<{
+				id: number;
+				belongs_to__application: { __id: number };
+			}>;
+			const devicesByApp = _.groupBy(
+				devices,
+				(d) => d.belongs_to__application.__id,
+			);
+
+			await Promise.all(
+				Object.keys(devicesByApp).map((appIdKey) => {
+					const appId = devicesByApp[appIdKey][0].id;
+
+					// delete service installs for devices currently with this app
+					actOnAppServiceInstalls(
```

Comment:
This is missing an `await`
```suggestion
					await actOnAppServiceInstalls(
```